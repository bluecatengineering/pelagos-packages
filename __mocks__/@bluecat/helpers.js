export const buildHighlighter = jest.fn();
export const smoothScroll = jest.fn();

export const getLogger = jest.fn(() => ({
	logError: jest.fn(),
	logInfo: jest.fn(),
	logDebug: jest.fn(),
}));
