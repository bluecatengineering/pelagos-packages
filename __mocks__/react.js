module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
