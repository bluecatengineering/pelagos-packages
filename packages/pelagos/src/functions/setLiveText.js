const getLiveElement = () => {
	let element = document.getElementById('liveText');
	if (!element) {
		element = document.createElement('div');
		element.id = 'liveText';
		element.className = 'assistive-text';
		element.setAttribute('aria-live', 'assertive');
		document.body.appendChild(element);
	}
	return element;
};

/**
 * Sets the specified text on an element marked as aria-live="assertive".
 * The element is created as required and reused in subsequent calls.
 * @param {string} text the text.
 */
export default (text) => (getLiveElement().textContent = text);
