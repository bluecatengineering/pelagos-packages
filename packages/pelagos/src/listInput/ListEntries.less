@import '../../less/breakpoints';
@import '../../less/spacing';
@import '../../less/utils';

.ListEntries {
	gap: @sp-08;
	margin: 0;
	padding: 0;
	list-style: none;

	&--column {
		display: flex;
		flex-direction: column;
	}

	&--grid {
		display: grid;
		grid-template-columns: 1fr;
		grid-auto-rows: min-content;

		.breakpoint(md, {grid-template-columns: repeat(2, 1fr)});
		.breakpoint(lg, {grid-template-columns: repeat(4, 1fr)});
		.breakpoint(xlg, {grid-template-columns: repeat(8, 1fr)});
	}
}

.ListEntries__item {
	display: flex;
	flex-direction: row;
	align-items: center;
	overflow: hidden;
}

.ListEntries__item--highlight {
	animation: highlight-fade 2s ease-out;
}

.ListEntries__icon {
	@icon-button-nested();
	font-size: 16px;
	padding: @sp-02;

	.ListEntries__item--highlight > & {
		color: var(--link-primary);
	}
}

.ListEntries__name {
	flex: 1;
	padding-left: @sp-04;
}
