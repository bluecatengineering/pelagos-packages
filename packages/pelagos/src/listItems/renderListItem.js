import ListItem from './ListItem';

/**
 * Renders a list item.
 * @param {string} item the item to render.
 * @param {string} [className] the class name(s).
 * @returns {JSX.Element}
 */
// eslint-disable-next-line react/display-name -- false positive, this is not a component
export default (item, className) => <ListItem key={item} item={item} className={className} />;
