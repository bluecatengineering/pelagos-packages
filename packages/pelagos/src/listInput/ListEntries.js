import {t} from '@bluecateng/l10n.macro';
import Close from '@carbon/icons-react/es/Close';
import Draggable from '@carbon/icons-react/es/Draggable';
import debounce from 'lodash-es/debounce';
import PropTypes from 'prop-types';
import {cloneElement, useCallback, useEffect, useMemo} from 'react';

import Layer from '../components/Layer';
import moveListItem from '../functions/moveListItem';
import scrollIntoView from '../functions/scrollIntoView';
import useReorder from '../hooks/useReorder';
import renderListItem from '../listItems/renderListItem';

import './ListEntries.less';

/** A list of entries. */
const ListEntries = ({
	id,
	className,
	highlightKey,
	list,
	reorderable,
	column,
	getItemKey,
	getItemName,
	renderItem,
	onReorder,
	onRemoveClick,
	onHighlightClear,
}) => {
	const getElementName = useCallback((element) => getItemName(list[+element.dataset.index]), [list, getItemName]);
	const updateList = useCallback(
		(fromIndex, toIndex) => onReorder(moveListItem(list, fromIndex, toIndex)),
		[list, onReorder]
	);
	const [reorderRef, liveRef] = useReorder(
		'.ListEntries__item',
		'.ListEntries__entry',
		list.length,
		getElementName,
		updateList
	);

	const clearHighlight = useMemo(() => onHighlightClear && debounce(onHighlightClear, 1000), [onHighlightClear]);

	const handleClick = useCallback(
		(event) => {
			const target = event.target.closest('button[data-index]');
			if (target) {
				const index = +target.dataset.index;
				const item = list[index];
				const name = getItemName(item);
				liveRef.current.textContent = t`${name} removed`;
				onRemoveClick(item, index);
			}
		},
		[list, getItemName, liveRef, onRemoveClick]
	);

	useEffect(() => {
		if (highlightKey) {
			const element = document.querySelector('.ListEntries__item--highlight');
			if (element) {
				scrollIntoView(element, {smooth: true, duration: 150});
			}
			clearHighlight();
		}
		return clearHighlight && clearHighlight.cancel;
	}, [highlightKey, clearHighlight]);

	const operationId = `${id}-operation`;
	return (
		<>
			<div className="sr-only" aria-live="polite" ref={liveRef} />
			{reorderable && (
				<div id={operationId} className="sr-only">
					{t`Press space bar to reorder`}
				</div>
			)}
			<ul
				ref={reorderRef}
				id={id}
				className={`ListEntries ListEntries--${column || reorderable ? 'column' : 'grid'}${className ? ` ${className}` : ''}`}
				onClick={handleClick}>
				{list.map((item, i) => {
					const name = getItemName(item);
					const element = renderItem ? renderItem(item) : renderListItem(name);
					let className = element.props.className;
					className = className ? `${className} ListEntries__name` : 'ListEntries__name';
					const itemKey = getItemKey(item, i);
					return (
						<Layer
							key={itemKey}
							as="li"
							className={`ListEntries__item${itemKey === highlightKey ? ' ListEntries__item--highlight' : ''}`}
							tabIndex={reorderable ? 0 : undefined}
							aria-describedby={reorderable ? operationId : undefined}
							data-testid="list-item"
							data-index={i}>
							{reorderable ? (
								<div className="ListEntries__entry draggable">
									<Draggable className="ListEntries__grip" />
									{cloneElement(element, {className})}
								</div>
							) : (
								cloneElement(element, {className})
							)}
							<button
								className="ListEntries__icon"
								type="button"
								aria-label={t`Remove ${name}`}
								data-testid="remove-item"
								data-index={i}>
								<Close />
							</button>
						</Layer>
					);
				})}
			</ul>
		</>
	);
};

ListEntries.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The key of the highlighted row. */
	highlightKey: PropTypes.string,
	/** The data for the list. */
	list: PropTypes.array,
	/** Whether items are reorderable. */
	reorderable: PropTypes.bool,
	/** Whether items are listed as columns. Set to true if reorderable. */
	column: PropTypes.bool,
	/** Function invoked to get each item's key. */
	getItemKey: PropTypes.func,
	/** Function invoked to get each item's name. */
	getItemName: PropTypes.func,
	/** Function invoked to render each list item. */
	renderItem: PropTypes.func,
	/** Function invoked when an item is reordered. */
	onReorder: PropTypes.func,
	/** Function invoked when the remove button is clicked. */
	onRemoveClick: PropTypes.func,
	/** Function invoked to clear the highlight key. */
	onHighlightClear: PropTypes.func,
};

export default ListEntries;
