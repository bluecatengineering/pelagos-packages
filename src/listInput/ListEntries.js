import React, {cloneElement, useCallback, useEffect, useMemo} from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash-es/debounce';
import {scrollIntoView} from '@bluecat/helpers';

import useButtonKeyHandler from '../hooks/useButtonKeyHandler';
import renderListItem from '../listItems/renderListItem';
import SvgIcon from '../components/SvgIcon';
import timesThin from '../icons/timesThin';
import useLiveText from '../hooks/useLiveText';
import __ from '../strings';

import './ListEntries.less';

/** A list of entries. */
const ListEntries = ({
	id,
	highlightKey,
	list,
	column,
	getItemKey,
	getItemName,
	renderItem,
	onRemoveClick,
	onHighlightClear,
}) => {
	const clearHighlight = useMemo(() => onHighlightClear && debounce(onHighlightClear, 1000), [onHighlightClear]);

	const setLiveText = useLiveText();

	const handleClick = useCallback(
		(event) => {
			const target = event.target.closest('[data-index]');
			if (target) {
				const index = +target.dataset.index;
				const item = list[index];
				setLiveText(__('OBJECT_REMOVED', {name: getItemName(item)}));
				onRemoveClick(item, index);
			}
		},
		[list, onRemoveClick, getItemName, setLiveText]
	);

	const handleKeyDown = useButtonKeyHandler();

	useEffect(() => {
		if (highlightKey) {
			const element = document.querySelector('.ListEntries__item--highlight');
			if (element) {
				scrollIntoView(element, {smooth: true, duration: 150});
			}
			clearHighlight();
		}
		return clearHighlight && clearHighlight.cancel;
	}, [highlightKey, clearHighlight]);

	return (
		<div id={id} className="ListEntries" role="list" onClick={handleClick} onKeyDown={handleKeyDown}>
			{list.map((item, i) => {
				const name = getItemName(item);
				const element = renderItem ? renderItem(item) : renderListItem(name);
				let className = element.props.className;
				className = className ? className + ' ListEntries__name' : 'ListEntries__name';
				const itemKey = getItemKey(item, i);
				return (
					<div
						key={itemKey}
						className={
							'ListEntries__item ' +
							(itemKey === highlightKey ? ' ListEntries__item--highlight' : '') +
							(column ? '' : ' ListEntries__gridItem')
						}
						role="listitem"
						data-bcn-id="list-item">
						<div
							data-bcn-id="remove-item"
							className="ListEntries__icon"
							tabIndex="0"
							role="button"
							aria-label={__('REMOVE_OBJECT', {name})}
							data-index={i}>
							<SvgIcon icon={timesThin} />
						</div>
						{cloneElement(element, {className})}
					</div>
				);
			})}
		</div>
	);
};

ListEntries.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The key of the highlighted row. */
	highlightKey: PropTypes.string,
	/** The data for the list. */
	list: PropTypes.array,
	/** Whether items are listed as columns. */
	column: PropTypes.bool,
	/** Function invoked to get each item's key. */
	getItemKey: PropTypes.func,
	/** Function invoked to get each item's name. */
	getItemName: PropTypes.func,
	/** Function invoked to render each list item. */
	renderItem: PropTypes.func,
	/** Function invoked when the remove button is clicked. */
	onRemoveClick: PropTypes.func,
	/** Function invoked to clear the highlight key. */
	onHighlightClear: PropTypes.func,
};

export default ListEntries;
