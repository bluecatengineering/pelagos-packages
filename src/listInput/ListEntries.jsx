import React, {cloneElement, useEffect, useMemo} from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash-es/debounce';
import {scrollIntoView} from '@bluecat/helpers';

import SvgIcon from '../components/SvgIcon';
import timesThin from '../icons/timesThin';
import __ from '../strings';

import './ListEntries.less';

const ListEntries = ({id, highlightKey, list, column, getItemKey, renderItem, onRemoveClick, onHighlightClear}) => {
	const clearHighlight = useMemo(() => onHighlightClear && debounce(onHighlightClear, 1000), [onHighlightClear]);

	useEffect(() => {
		if (highlightKey) {
			const element = document.querySelector('.ListEntries__item--highlight');
			if (element) {
				scrollIntoView(element, {smooth: true, duration: 150});
			}
			clearHighlight();
		}
		return clearHighlight && clearHighlight.cancel;
	}, [highlightKey, clearHighlight]);

	return (
		<div id={id} className="ListEntries">
			{list.map((item, i) => {
				const element = renderItem(item);
				let className = element.props.className;
				className = className ? className + ' ListEntries__name' : 'ListEntries__name';
				const itemKey = getItemKey(item, i);
				return (
					<div
						data-bcn-id="list-item"
						className={
							'ListEntries__item ' +
							(itemKey === highlightKey ? ' ListEntries__item--highlight' : '') +
							(column ? '' : ' ListEntries__gridItem')
						}
						key={itemKey}>
						<div
							data-bcn-id="remove-item"
							className="ListEntries__icon"
							role="button"
							aria-label={__('REMOVE')}
							onClick={() => onRemoveClick(item, i)}>
							<SvgIcon icon={timesThin} />
						</div>
						{cloneElement(element, {className})}
					</div>
				);
			})}
		</div>
	);
};

ListEntries.propTypes = {
	id: PropTypes.string,
	highlightKey: PropTypes.string,
	list: PropTypes.array,
	column: PropTypes.bool,
	getItemKey: PropTypes.func,
	renderItem: PropTypes.func,
	onRemoveClick: PropTypes.func,
	onHighlightClear: PropTypes.func,
};

export default ListEntries;
