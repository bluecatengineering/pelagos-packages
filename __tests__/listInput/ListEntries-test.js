import {useEffect} from 'react';
import {shallow} from 'enzyme';
import debounce from 'lodash-es/debounce';
import {scrollIntoView} from '@bluecat/helpers';

import ListEntries from '../../src/listInput/ListEntries';
import renderListItem from '../../src/listItems/renderListItem';
import setLiveText from '../../src/functions/setLiveText';

jest.unmock('../../src/listInput/ListEntries');
jest.unmock('../../src/strings');

jest.mock('lodash-es/debounce', () => jest.fn((f) => ((f.cancel = jest.fn()), f)));

const getId = (i) => i.id;
const getName = (i) => i.name;

const querySelector = jest.fn();
global.document = {querySelector};

describe('ListEntries', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<ListEntries
					id="test"
					list={[{id: '0', name: 'test'}]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when renderItem is not set', () => {
			renderListItem.mockReturnValue(<div>test</div>);
			const wrapper = shallow(
				<ListEntries id="test" list={[{id: '0', name: 'test'}]} getItemKey={getId} getItemName={getName} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
			expect(renderListItem.mock.calls).toEqual([['test']]);
		});

		it('renders expected elements when the item has className', () => {
			const wrapper = shallow(
				<ListEntries
					id="test"
					list={[{id: '0', name: 'test'}]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div className="TestClass">{i.name}</div>}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when highlightKey is set', () => {
			const wrapper = shallow(
				<ListEntries
					id="test"
					highlightKey="0"
					list={[{id: '0', name: 'test'}]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when column is set', () => {
			const wrapper = shallow(
				<ListEntries
					id="test"
					list={[{id: '0', name: 'test'}]}
					column
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onRemoveClick when the remove button is clicked', () => {
			const onRemoveClick = jest.fn();
			const item = {id: '0', name: 'test'};
			const wrapper = shallow(
				<ListEntries
					id="test"
					list={[item]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
					onRemoveClick={onRemoveClick}
				/>
			);
			wrapper.find('[role="list"]').simulate('click', {target: {closest: () => ({dataset: {index: '0'}})}});
			expect(onRemoveClick.mock.calls).toEqual([[item, 0]]);
			expect(setLiveText.mock.calls).toEqual([['test removed']]);
		});

		it('does not call onRemoveClick when the list is clicked outside', () => {
			const onRemoveClick = jest.fn();
			const item = {id: '0', name: 'test'};
			const wrapper = shallow(
				<ListEntries
					id="test"
					list={[item]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
					onRemoveClick={onRemoveClick}
				/>
			);
			wrapper.find('[role="list"]').simulate('click', {target: {closest: () => null}});
			expect(onRemoveClick.mock.calls).toEqual([]);
		});

		it('calls scrollIntoView when highlightKey is set and the element is found', () => {
			const onHighlightClear = jest.fn();
			const element = {};
			querySelector.mockReturnValue(element);
			shallow(
				<ListEntries
					id="test"
					highlightKey="0"
					list={[{id: '0', name: 'test'}]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
					onHighlightClear={onHighlightClear}
				/>
			);
			expect(useEffect.mock.calls).toEqual([[expect.any(Function), ['0', onHighlightClear]]]);
			const debounced = debounce.mock.results[0].value;
			expect(debounced).toBe(onHighlightClear);

			expect(useEffect.mock.calls[0][0]()).toBe(debounced.cancel);
			expect(querySelector.mock.calls).toEqual([['.ListEntries__item--highlight']]);
			expect(scrollIntoView.mock.calls).toEqual([[element, {smooth: true, duration: 150}]]);
			expect(onHighlightClear).toHaveBeenCalledTimes(1);
		});

		it('does not call scrollIntoView when highlightKey is set and the element is not found', () => {
			const onHighlightClear = jest.fn();
			querySelector.mockReturnValue(null);
			shallow(
				<ListEntries
					id="test"
					highlightKey="1"
					list={[{id: '0', name: 'test'}]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
					onHighlightClear={onHighlightClear}
				/>
			);
			useEffect.mock.calls[0][0]();
			expect(querySelector.mock.calls).toEqual([['.ListEntries__item--highlight']]);
			expect(scrollIntoView).not.toHaveBeenCalled();
			expect(onHighlightClear).toHaveBeenCalledTimes(1);
		});

		it('does not call querySelector when highlightKey is not set', () => {
			const onHighlightClear = jest.fn();
			shallow(
				<ListEntries
					id="test"
					list={[{id: '0', name: 'test'}]}
					getItemKey={getId}
					getItemName={getName}
					renderItem={(i) => <div>{i.name}</div>}
					onHighlightClear={onHighlightClear}
				/>
			);
			useEffect.mock.calls[0][0]();
			expect(querySelector).not.toHaveBeenCalled();
			expect(onHighlightClear).not.toHaveBeenCalled();
		});
	});
});
