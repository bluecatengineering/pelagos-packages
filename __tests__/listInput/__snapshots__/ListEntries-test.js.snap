// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListEntries rendering renders expected elements 1`] = `
<div
  className="ListEntries"
  id="test"
>
  <div
    className="ListEntries__item  ListEntries__gridItem"
    data-bcn-id="list-item"
  >
    <div
      aria-label="Remove"
      className="ListEntries__icon"
      data-bcn-id="remove-item"
      onClick={[Function]}
      role="button"
    >
      <SvgIcon
        icon={
          Object {
            "icon": Array [],
          }
        }
      />
    </div>
    <div
      className="ListEntries__name"
    >
      test
    </div>
  </div>
</div>
`;

exports[`ListEntries rendering renders expected elements when column is set 1`] = `
<div
  className="ListEntries"
  id="test"
>
  <div
    className="ListEntries__item "
    data-bcn-id="list-item"
  >
    <div
      aria-label="Remove"
      className="ListEntries__icon"
      data-bcn-id="remove-item"
      onClick={[Function]}
      role="button"
    >
      <SvgIcon
        icon={
          Object {
            "icon": Array [],
          }
        }
      />
    </div>
    <div
      className="ListEntries__name"
    >
      test
    </div>
  </div>
</div>
`;

exports[`ListEntries rendering renders expected elements when highlightKey is set 1`] = `
<div
  className="ListEntries"
  id="test"
>
  <div
    className="ListEntries__item  ListEntries__item--highlight ListEntries__gridItem"
    data-bcn-id="list-item"
  >
    <div
      aria-label="Remove"
      className="ListEntries__icon"
      data-bcn-id="remove-item"
      onClick={[Function]}
      role="button"
    >
      <SvgIcon
        icon={
          Object {
            "icon": Array [],
          }
        }
      />
    </div>
    <div
      className="ListEntries__name"
    >
      test
    </div>
  </div>
</div>
`;

exports[`ListEntries rendering renders expected elements when the item has className 1`] = `
<div
  className="ListEntries"
  id="test"
>
  <div
    className="ListEntries__item  ListEntries__gridItem"
    data-bcn-id="list-item"
  >
    <div
      aria-label="Remove"
      className="ListEntries__icon"
      data-bcn-id="remove-item"
      onClick={[Function]}
      role="button"
    >
      <SvgIcon
        icon={
          Object {
            "icon": Array [],
          }
        }
      />
    </div>
    <div
      className="TestClass ListEntries__name"
    >
      test
    </div>
  </div>
</div>
`;
